keps/sig-auth/20190607-certificates-api.md

```diff
@@ -0,0 +1,266 @@
+  // request approval state, currently Approved or Denied.
+  Type RequestConditionType
+  // brief reason for the request state
+  Reason string
+  // human readable message with details about the request state
+  Message string
+}
+
+type RequestConditionType string
+
+// These are the possible conditions for a certificate request.
+const (
+  CertificateApproved RequestConditionType = "Approved"
+  CertificateDenied   RequestConditionType = "Denied"
+)
+```
+
+### Manual CSR Approval With Kubectl
+
+A Kubernetes administrator (with appropriate permissions) can manually approve
+(or deny) Certificate Signing Requests by using the `kubectl certificate
+approve` and `kubectl certificate deny` commands.
+
+### Automatic CSR Approval Implementations
+
+The kube-controller-manager ships with an in-built
+[approver](https://github.com/kubernetes/kubernetes/blob/32ec6c212ec9415f604ffc1f4c1f29b782968ff1/pkg/controller/certificates/approver/sarapprove.go)
+for Kubelet TLS Bootstrap that delegates various permissions on CSRs for node
+credentials to authorization. It does this by posting subject access reviews to
+the API server. It punts on TLS certificates for server authentication of the
+Kubelet API because verifying IP SANs for Kubelets in a generic way poses
+challenges.
+
+The GCP controller manager replaces the in-built approver with an
+[approver](https://github.com/kubernetes/cloud-provider-gcp/blob/08fa1e3260ffb267682762e24ba93692000e3be8/cmd/gcp-controller-manager/csr_approver.go)
+that handles TLS certificates for server authentication and support for
+verification of CSRs with proof of control of GCE vTPM.
+
+An external project, [kapprover](https://github.com/coreos/kapprover), does
+policy based approval of kubelet CSRs.
+
+<!--
+TODO: various things use the certificates API for stuff other than Kubelet TLS
+certficates. For example:
+
+* Istio
+* CockroachDB
+
+Investigate these usages and report back.
+-->
+
+### Automatic Signer Implementations
+
+The kube-controller-manager ships with an in-built
+[signer](https://github.com/kubernetes/kubernetes/blob/32ec6c212ec9415f604ffc1f4c1f29b782968ff1/pkg/controller/certificates/signer/cfssl_signer.go)
+that signs all approved certificates with a local key. This key is generally
+configured to be the key of the cluster's root certificate authority, but is not
+required to be.
+
+The GCP controller manager implements a
+[signer](https://github.com/kubernetes/cloud-provider-gcp/blob/08fa1e3260ffb267682762e24ba93692000e3be8/cmd/gcp-controller-manager/csr_signer.go)
+that uses a webhook to sign all approved CSRs. This allows the root certificate
+authority secret material to be stored and maintained outside of the Kubernetes
+control plane.
+
+<!--
+### Certificates API Clients
+
+TODO: talk about what's provided by k8s.io/client-go and what other people have
+built, e.g.
+[certificate-init-container](https://github.com/kelseyhightower/certificate-init-container).
+-->
+
+## Graduation Criteria
```

Comment:
We have to define which flows are required as part of conformance in here.  Ie are you required to issue Kubelet client and server csrs to be conformant, what bounds exist, etc.

I want this backfilled so we can iterate so I’m willing to wait for a follow up if you can complete the mechanical comments others have left.

Reply:
Yup. I'm hoping this KEP provides a place to have those discussions. This KEP started with a list of known issues but it got dropped to streamline initial review. My list is here but I am looking for help brainstorming.

- API shape issues
  - Requires requester to know all the info about the end certificate.
  - No support for higher-level requests (“give me a node client cert”, “give me a serving cert for a pod backing a service”).
  - Requested certificate attributes are split unthoughtfully between encoded CSR and fields in request spec which create difference in semantics.
- Approval flow issues
  - Approvers cannot add components to requests (e.g. SANs, usages, etc)
  - Approvers cannot be limited in what certificates they approve (i.e. subdivision of approvers)
  - Some approval flows that grant client certs trusted by the API server allow for arbitrary credential lifetime extension.
- Signing flow issues
  - No support for multiple signers to interact (or approver to indicate what signer should be used)
- No guarantees on issued certificates
  - No guarantee all requested extensions/SANs are issued
  - No guarantee on who trusts the certificate (e.g. is a client certificate going to work against the API?) 
- Certain user journeys that we care about in core remain challenging, such as securing a webhook or extension API running on the cluster. We need to decide the scope of the approvers shipped in core and understand which journeys should be supported out of the box.